bolts/fair/models/fair_mixup.py

```diff
@@ -0,0 +1,307 @@
+        return self._inference_step(batch=batch, stage="test")
+
+    @implements(pl.LightningModule)
+    def configure_optimizers(
+        self,
+    ) -> tuple[list[optim.Optimizer], list[Mapping[str, LRScheduler | int | SchedInterval]]]:
+        opt = optim.AdamW(
+            self.parameters(),
+            lr=self.learning_rate,
+            weight_decay=self.weight_decay,
+        )
+        sched = {
+            "scheduler": CosineAnnealingWarmRestarts(
+                optimizer=opt, T_0=self.lr_initial_restart, T_mult=self.lr_restart_mult
+            ),
+            "interval": self.lr_sched_interval,
+            "frequency": self.lr_sched_freq,
+        }
+        return [opt], [sched]
+
+    @staticmethod
+    def mixup_data(
+        *,
+        batch: DataBatch,
+        device: torch.device,
+        mix_lambda: Optional[float],
+        alpha: float,
+        fairness: FairnessType,
+    ) -> tuple[Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, dict[str, float]]:
+        '''Returns mixed inputs, pairs of targets, and lambda'''
+        lam = (
+            Beta(
+                torch.tensor([alpha]).to(device),
+                torch.tensor([alpha]).to(device)
+                # Potentially change alpha from a=1.0 to account for class imbalance?
+            ).sample()
+            if mix_lambda is None
+            else torch.tensor([mix_lambda]).to(device)
+        )
+
+        batch_x_s0 = batch.x[batch.s.view(-1) == 0].to(device)
+        batch_x_s1 = batch.x[batch.s.view(-1) == 1].to(device)
+        batch_y_s0 = batch.y[batch.s.view(-1) == 0].to(device)
+        batch_y_s1 = batch.y[batch.s.view(-1) == 1].to(device)
+
+        batch_x_s0_y0 = batch.x[(batch.s.view(-1) == 0) & (batch.y.view(-1) == 0)].to(device)
+        batch_x_s1_y0 = batch.x[(batch.s.view(-1) == 1) & (batch.y.view(-1) == 0)].to(device)
+        batch_x_s0_y1 = batch.x[(batch.s.view(-1) == 0) & (batch.y.view(-1) == 1)].to(device)
+        batch_x_s1_y1 = batch.x[(batch.s.view(-1) == 1) & (batch.y.view(-1) == 1)].to(device)
+
+        batch_s_s0_y0 = batch.s[(batch.s.view(-1) == 0) & (batch.y.view(-1) == 0)].to(device)
+        batch_s_s1_y0 = batch.s[(batch.s.view(-1) == 1) & (batch.y.view(-1) == 0)].to(device)
+        batch_s_s0_y1 = batch.s[(batch.s.view(-1) == 0) & (batch.y.view(-1) == 1)].to(device)
+        batch_s_s1_y1 = batch.s[(batch.s.view(-1) == 1) & (batch.y.view(-1) == 1)].to(device)
+
+        xal = []
+        xbl = []
+        sal = []
+        sbl = []
+        yal = []
+        ybl = []
+
+        for x_a, s_a, y_a in zip(batch.x, batch.s, batch.y):
+            xal.append(x_a)
+            sal.append(s_a.unsqueeze(-1).float())
+            yal.append(y_a.unsqueeze(-1).float())
+            if fairness is FairnessType.DP:
+                idx = torch.randint(eval(f"batch_x_s{1-int(s_a)}").size(0), (1,))
+                x_b = eval(f"batch_x_s{1-int(s_a)}")[idx, :].squeeze(0)
```

Comment:
`eval()` is very slow and also the opposite of safe. I think you should just use a dictionary `batch_x_s[0]` etc. this would also allow using loops for the masking code above.